Cython/Utility/TypeConversion.c

```diff
@@ -286,20 +291,61 @@ static CYTHON_INLINE const char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_
         return PyByteArray_AS_STRING(o);
 #else
         *length = PyByteArray_Size(o);
-        if (*length == -1) return NULL;
+        if (*length == -1) goto error;
         return PyByteArray_AsString(o);
 #endif
 
-    } else
-    {
+    } else if (PyBytes_Check(o)) {
         char* result;
         int r = PyBytes_AsStringAndSize(o, &result, length);
         if (unlikely(r < 0)) {
-            return NULL;
+            goto error;
         } else {
             return result;
         }
+    } else {
+        char* result;
+        int r;
+#if !CYTHON_COMPILING_IN_LIMITED_API || __PYX_LIMITED_VERSION_HEX >= 0x030B0000
+        // Check that `o`:
+        // * Supports the Python Buffer Protocol
+        // * There is no need to keep the buffer around (as we cannot return it)
+        PyBufferProcs *pb = Py_TYPE(o)->tp_as_buffer;
+        if (pb == NULL || pb->bf_getbuffer == NULL || pb->bf_releasebuffer != NULL) {
+            goto error;
+        }
+
+        // Try to acquire buffer from `o`
+        Py_buffer view;
+        r = PyObject_GetBuffer(o, &view, PyBUF_SIMPLE);
+        if (unlikely(r < 0)) {
+            goto error;
+        } else {
+            result = (char*) view.buf;
+            *length = view.len;
+            PyBuffer_Release(&view);  // Release to clean up buffer (decref object `o`)
+            return result;
+        }
+#else
+        // Fallback for the Limited API on Python pre-3.11
+        r = PyArg_Parse(o, "y#", &result, &length);
+        if (unlikely(r < 0)) {
+            goto error;
+        } else {
+            return result;
+        }
+#endif
     }
+
+    error:
+        // Clear existing errors and return our standardized one
+        PyErr_Clear();
+        __Pyx_TypeName result_type_name = __Pyx_PyType_GetFullyQualifiedName(Py_TYPE(result));
+        PyErr_Format(PyExc_TypeError,
+            "a bytes-like object is required, not '" __Pyx_FMT_TYPENAME "'",
+            result_type_name);
+        *length = -1;
+        return NULL;
```

Comment:
Added a generic error for all cases to use here. Please let me know what you think about this and if there are things we want to improve about it